### PR TITLE
Generate documented __init__ methods and better document custom __init__ methods

### LINKED ISSUE #3423: `**kwargs` in constructors makes archetype creation more difficult
**Describe the annoyance**
When constructors of archetypes take `**kwargs` it becomes harder to create an archetype. It becomes unclear what arguments are supported. For example, `class_ids`, `colors`, etc. are not visible for `Boxes2D`, and `Boxes3D`.

kwargs

**Expected behavior**
Show all arguments.

**Your goals**
Ideally understand how to use the archetype from just looking at the signature.

### LINKED ISSUE #3433: Archetype doc strings should explain parameters
Currently, the docstrings for archetypes in python don't include documentation of the parameters of the init methods. The types are not descriptive enough which leaves the API's very hard to understand.

For example, this is quite hard to use:
Screenshot 2023-09-25 at 09 19 55

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes #3423
* Fixes #3433

---

* We now always emit `__init__` methods in python alongside with a doc string.
* Improved docstring for manual `__init__` methods
* fixed lots of type issues that bubbled up by doing the above

LSP is now very happy with me on pretty much all types:

manual `__init__`:
image

generated:
image

mouse hovering the type itself:
image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3480) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3480)
- [Docs preview](https://rerun.io/preview/3e2e42f9097e5586026f23cd77b6adbb061e7c69/docs) 
- [Examples preview](https://rerun.io/preview/3e2e42f9097e5586026f23cd77b6adbb061e7c69/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: examples/python/notebook/blueprint.ipynb [hunk 1]
```diff
@@ -1,128 +1,132 @@
 {
-    "cells": [
```
Comment (Wumpf) [inline examples/python/notebook/blueprint.ipynb:2]: looks like my code formatter has a different opinion here than the last person who touched it

File: crates/re_types_builder/src/codegen/python.rs [hunk 2]
```diff
@@ -1491,6 +1491,156 @@ fn quote_arrow_support_from_obj(
...
+    };
+
+    if field.is_nullable {
+        format!("{}: {} | None = None", field.name, type_annotation)
+    } else {
+        format!("{}: {}", field.name, type_annotation)
+    }
+}
+
+fn quote_init_method(obj: &Object, ext_class: &ExtensionClass, objects: &Objects) -> String {
```
Comment (Wumpf) [inline crates/re_types_builder/src/codegen/python.rs:1543]: this code can surely golfed down so something more compact, but I already spent to much time on that aspect. If someone sees easy shortcuts that would be appreciated though.

File: examples/python/notebook/blueprint.ipynb (hunk 1 above)
Comment (teh-cmc) [reply examples/python/notebook/blueprint.ipynb:2]: same for me, so I guess this is the correct one

File: rerun_py/rerun_sdk/rerun/components/tensor_data.py [hunk 3]
```diff
@@ -5,17 +5,25 @@
...
 from .. import datatypes
 from .._baseclasses import ComponentBatchMixin
 
 __all__ = ["TensorData", "TensorDataBatch", "TensorDataType"]
 
 
+@define(init=False)
 class TensorData(datatypes.TensorData):
-    # You can define your own __init__ function as a member of TensorDataExt in tensor_data_ext.py
+    def __init__(self: Any, shape: datatypes.TensorDimensionArrayLike, buffer: datatypes.TensorBufferLike):
```
Comment (jleibs) [inline rerun_py/rerun_sdk/rerun/components/tensor_data.py:20]: This is is not correct. Previously delegating components would dispatch to the constructor of the Datatype, which, importantly, would in turn dispatch to the extension type.

In this case it meant that `components.TensorData` would use the constructor override from `datatypes.TensorDataExt` automatically if it didn't provide its own TensorDataExt.

Comment (Wumpf) [reply rerun_py/rerun_sdk/rerun/components/tensor_data.py:20]: ah, so for delegating ones we shouldn't call the attribute init but the regular init then!

Comment (Wumpf) [reply rerun_py/rerun_sdk/rerun/components/tensor_data.py:20]: .. or would leaving it out completely work? That be best ofc!

Comment (Wumpf) [reply rerun_py/rerun_sdk/rerun/components/tensor_data.py:20]: yes. leaving it out should be the right thing to do + not putting `@define`
This should make quite a few things easier in the codegen actually!

Comment (jleibs) [reply rerun_py/rerun_sdk/rerun/components/tensor_data.py:20]: I was surprised the unit-tests didn't catch this but apparently we were only testing the datatype and not the component: https://github.com/rerun-io/rerun/pull/3486

**Comment by teh-cmc** [comment]
Based on the comments above, I'm gonna guess this is _not_ ready for review?

**Comment by Wumpf** [comment]
correct. converted to draft

**Comment by Wumpf** [comment]
was an easy fix but also didn't simplfy the code as much as I hoped :)

File: rerun_py/rerun_sdk/rerun/components/tensor_data.py [hunk 4]
```diff
@@ -12,8 +12,6 @@
 
 
 class TensorData(datatypes.TensorData):
-    # You can define your own __init__ function as a member of TensorDataExt in tensor_data_ext.py
```
Comment (jleibs) [inline rerun_py/rerun_sdk/rerun/components/tensor_data.py:15]: Nit: this comment should still be here.  There may be a valid reason to implement an Extension for a delegating component (even if we don't yet).

File: rerun_py/rerun_sdk/rerun/archetypes/pinhole_ext.py [hunk 5]
```diff
@@ -78,6 +22,62 @@ def __init__(
...
+        ----------
+        image_from_camera:
+            Row-major intrinsics matrix for projecting from camera space to image space.
+            The first two axes are X=Right and Y=Down, respectively.
+            Projection is done along the positive third (Z=Forward) axis.
+            This can be specified _instead_ of `focal_length` and `principal_point`.
+        resolution:
+            Pixel resolution (usually integers) of child image space. Width and height.
+            `image_from_camera` projects onto the space spanned by `(0,0)` and `resolution - 1`.
+        camera_xyz:
```
Comment (jleibs) [inline rerun_py/rerun_sdk/rerun/archetypes/pinhole_ext.py:38]: TODO: Would be nice to patch these during code-gen if the parameter name matches.

**Review by jleibs — APPROVED** [review]

**Pull request merged**